src/daemon/distribution.c

```diff
@@ -0,0 +1,402 @@
+#include <math.h>
+#include <pthread.h>
+
+typedef struct {
+  double max_boundary;
+  uint64_t counter;
+} bucket_t;
+
+struct distribution_s {
+  bucket_t *buckets;
+  double sum_gauges;
+  size_t num_buckets;
+  pthread_mutex_t mutex;
+};
+
+double *distribution_get_buckets_boundaries(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  double *boundaries = calloc(d->num_buckets, sizeof(double));
+
+  if (boundaries == NULL) {
+    return NULL;
+  }
+
+  for (size_t i = 0; i < d->num_buckets; ++i) {
+    boundaries[i] = d->buckets[i].max_boundary;
+  }
+
+  return boundaries;
+}
+
+uint64_t *distribution_get_buckets_counters(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  uint64_t *counters = calloc(d->num_buckets, sizeof(uint64_t));
+
+  if (counters == NULL) {
+    return NULL;
+  }
+
+  pthread_mutex_lock(&d->mutex);
+  for (size_t i = 0; i < d->num_buckets; ++i) {
+    counters[i] = d->buckets[i].counter;
+  }
+  pthread_mutex_unlock(&d->mutex);
+
+  return counters;
+}
+
+size_t distribution_get_num_buckets(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return 0;
+  }
+
+  return d->num_buckets;
+}
+
+double distribution_get_sum_gauges(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return NAN;
+  }
+
+  return d->sum_gauges;
+}
+
+bool distribution_check_equal(distribution_t *d1, distribution_t *d2) {
+  if ((d1 == NULL && d2 != NULL) || (d1 != NULL && d2 == NULL)) {
+    return false;
+  }
+
+  /* thanks to previous condition we know that if d1 is NULL then d2 is NULL too
+   */
+  if (d1 == NULL) {
+    return true;
+  }
+
+  pthread_mutex_lock(&d1->mutex);
+  pthread_mutex_lock(&d2->mutex);
+
+  if (d1->sum_gauges != d2->sum_gauges) {
+    return false;
+  }
+
+  if (d1->num_buckets != d2->num_buckets) {
+    return false;
+  }
+
+  for (size_t i = 0; i < d1->num_buckets; ++i) {
+    if (d1->buckets[i].max_boundary != d2->buckets[i].max_boundary ||
+        d1->buckets[i].counter != d2->buckets[i].counter) {
+      return false;
+    }
+  }
+
+  pthread_mutex_unlock(&d2->mutex);
+  pthread_mutex_unlock(&d1->mutex);
+
+  return true;
+}
+
+static bucket_t *bucket_new_linear(size_t num_buckets, double size) {
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = (double)(i + 1) * size;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_exponential(size_t num_buckets, double base,
+                                        double factor) {
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  double multiplier = 1.0;
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = factor * multiplier;
+    multiplier *= base;
```

Comment:
I would also recommend to use pow function. As we found out, it's much more precise than multiplication many times

Reply:
I didn't use the pow function to avoid calculating boundary for every bucket in logarithm time, which could end with num_buckets multiply by log(num_buckets) time complexity for distribution_new_exponential function. And doing it in the for loop ends with linear time complexity.